working/1847 - FinalizationRegistry/proposal.md

```diff
@@ -0,0 +1,411 @@
+# Proposal for adding `FinalizationRegistry` and `WeakRef` to core libraries.
+
+Author: vegorov@google.com<br>Version: 0.1
+
+## Background
+
+This proposal includes two interconnected pieces of functionality:
+
+- an ability to create _weak references_ (`WeakRef`);
+- an ability to associate finalization actions with objects
+  (`FinalizationRegistry`).
+
+### Weak references
+
+_Weak reference_ is an object which holds to its _target_ in a manner that does
+not prevent runtime system from reclaiming the referenced object when it becomes
+_weakly reachable_ or, in other words, reachable only through weak references.
+When an object is deemed _weakly reachable_ all `WeakRef` objects referencing it
+are cleared and the object is eligible for reclamation.
```

Comment:
A data point: JS moved away from defining reachability as it makes discussions of what optimization are allowed harder. The usual "as-if" reasoning for e.g. dead variable elimination breaks down because GC is now observable.

We ended up defining it in terms of [possible future executions](https://tc39.es/ecma262/multipage/executable-code-and-execution-contexts.html#sec-liveness).

Reply:
That's a very reasonable and functional  definition of when an object is garbage: When no future expression evaluation not relying on weak references can end up evaluating to the object. At that point, clearing the weak references will definitely sever all ties to the object.